File: GuardianMonsters/battle/src/de/limbusdev/guardianmonsters/battle/AbilityMedia.java
package de.limbusdev.guardianmonsters.battle;

import de.limbusdev.guardianmonsters.guardians.abilities.Ability;
import de.limbusdev.guardianmonsters.media.SFXType;

/**
 * AbilityMedia
 *
 * Holds information about which animation and sound to use for an {@link Ability}.
 *
 * @author Georg Eckert 2017
 */
public class AbilityMedia
{
    private final String name;
    private final int sfxIndex;
    private final SFXType sfxType;
    private final AnimationType animationType;

    public AbilityMedia(String name, int sfxIndex, SFXType sfxType, AnimationType animationType)
    {
        this.name = name;
        this.sfxIndex = sfxIndex;
        this.sfxType = sfxType;
        this.animationType = animationType;
    }

    public String getName()
    {
        return name;
    }

    public int getSfxIndex()
    {
        return sfxIndex;
    }

    public SFXType getSfxType()
    {
        return sfxType;
    }

    public AnimationType getAnimationType()
    {
        return animationType;
    }

    @Override
    public String toString()
    {
        return "Media for Ability " + name + ": SFX " + sfxIndex + " of type " + sfxType + ", Anim: " + animationType;
    }

    @Override
    public boolean equals(Object obj)
    {
        if(!(obj instanceof AbilityMedia)) return false;
        AbilityMedia other = (AbilityMedia) obj;
        if(other.getSfxIndex() == other.getSfxIndex() &&
            other.sfxType.equals(sfxType) &&
            other.getName().equals(name) &&
            other.getAnimationType().equals(animationType))
        {
            return true;
        }
        return super.equals(obj);
    }
}

File: GuardianMonsters/battle/src/de/limbusdev/guardianmonsters/battle/AbilityMedia.kt
package de.limbusdev.guardianmonsters.battle

import de.limbusdev.guardianmonsters.guardians.abilities.Ability
import de.limbusdev.guardianmonsters.media.SFXType

/**
 * AbilityMedia
 *
 * Holds information about which animation and sound to use for an [Ability].
 *
 * @author Georg Eckert 2017
 */
class AbilityMedia
(
        val name: String,
        val sfxIndex: Int,
        val sfxType: SFXType,
        val animationType: AnimationType
) {
    override fun toString(): String
    {
        return "Media for Ability $name: SFX $sfxIndex of type $sfxType, Anim: $animationType"
    }

    override fun equals(other: Any?): Boolean
    {
        if(this === other)                  return true
        if(javaClass != other?.javaClass)   return false

        other as AbilityMedia

        if(name != other.name)                      return false
        if(sfxIndex != other.sfxIndex)              return false
        if(sfxType != other.sfxType)                return false
        if(animationType != other.animationType)    return false

        return true
    }

    override fun hashCode(): Int
    {
        var result = name.hashCode()
        result = 31 * result + sfxIndex
        result = 31 * result + sfxType.hashCode()
        result = 31 * result + animationType.hashCode()
        return result
    }
}
